### PR TITLE
Unicode and Percent Encoding

### PR DESCRIPTION
**Pull request by kelvinqian00**
- Add support for Unicode (Basic Multilingual Plane) characters in variables, blank nodes, and prefixed IRIs.
- Add support for percent encoding for prefixed IRIs.
- Ensure that language tags are alphanumeric strings, as per the spec.
- Optimize validation (in Clojure) using low-level Java constructs.

Benching the results of Clojure optimization:
```clojure
(require '[com.yetanalytics.flint.axiom.impl.validation :as v])

;; Before: Regex WITHOUT Unicode
(crit/quick-bench
 (re-matches #"\?\w+" (name '?foo)))
;; =>  98.5 ns

;; During: Regex WITH Unicode
(crit/quick-bench
 (re-matches v/var-regex (name '?foo)))
;; => 276.9 ns

;; After: New function WITH Unicode
(crit/quick-bench
 (v/valid-var-symbol? '?foo))
;; =>  67.3 ns
```
On longer strings the difference between regex and custom Unicode validation become more apparent:
```clojure
(def v '?supercalifragilisticexpialidocious)

(crit/quick-bench
 (re-matches #"\?\w+" (name v)))
;; => 335.6 ns

(crit/quick-bench
 (re-matches v/var-regex (name v)))
;; => 2522.1 ns

(crit/quick-bench
 (v/valid-var-symbol? v))
;; => 348.4 ns
```

### CODE REVIEW

File: src/main/com/yetanalytics/flint/axiom/impl/validation.cljc [hunk 1]
```diff
@@ -1,55 +1,561 @@
...
 
-(def prefix-iri-ns-regex
-  #"[A-Za-z_]([\w\-\.]*[\w\-])?")
+(def ^:private unicode-range-no-digits
+  [[(char->int \A) (char->int \Z)]
+   [(char->int \a) (char->int \z)]
+   [0x00C0 0x00D6] ; * Latin letters with diacritics + spacing modifiers
+   [0x00D8 0x00F6] ;   Excludes multiplication + division signs
+   [0x0370 0x037D] ; * Various alphabets (e.g. Greek, Cyrillic, Arabic)
+   [0x037F 0x1FFF] ;   Excludes Greek question mark
```
Comment (milt) [inline src/main/com/yetanalytics/flint/axiom/impl/validation.cljc:28]: Probably for the best; lol

File: src/test/com/yetanalytics/flint/spec/axiom_test.cljc [hunk 2]
```diff
@@ -1,85 +1,256 @@
...
+        "소녀시대"
+        "少女时代"
+        "少女時代"
+        "しょうじょじだい")))
+
+(def ^:private biang-biang-noodles
+  "Test string that includes supplemental CJK characters."
+  (str (char 0xd883) (char 0xdedd) ; Java only supports UTF-16 => surrogate pairs
+       (char 0xd883) (char 0xdedd)
+       "面"))
```
Comment (milt) [inline src/test/com/yetanalytics/flint/spec/axiom_test.cljc:111]: glorious

File: src/test/com/yetanalytics/flint/spec/axiom_test.cljc [hunk 3]
```diff
@@ -1,85 +1,256 @@
...
+      {:en-US "foo"}
+      {:en-80 "bar"})
+    (are [x] (not (v/valid-lang-map-literal? x))
+      {:-en "foo"}
+      {:-en-US "foo"}
+      {:en- "foo"}
+      {:en-US- "foo"}
+      {:en--US "foo"}
+      {:en-! "foo"}
+      {:zh-你好 "foo"})))
```
Comment (milt) [inline src/test/com/yetanalytics/flint/spec/axiom_test.cljc:225]: Showing all kinds of holes in GitHub's Clojure syntax highlighting here!

**Review by milt — APPROVED** [review]
I love this, not just for the obvious performance advantage but also the excellent comments and testing of a domain that can be really obtuse.

File: src/main/com/yetanalytics/flint/axiom/impl/validation.cljc (hunk 1 above)
Comment (kelvinqian00) [reply src/main/com/yetanalytics/flint/axiom/impl/validation.cljc:28]: It's actually explicitly forbidden by the spec: https://www.w3.org/TR/sparql11-query/#rPN_CHARS_BASE

Comment (milt) [reply src/main/com/yetanalytics/flint/axiom/impl/validation.cljc:28]: As it should be. Related: https://twitter.com/benbjohnson/status/533848879423578112

**Pull request merged**